tachyon/crypto/commitments/kzg/shplonk_unittest.cc

```diff
@@ -1,17 +1,138 @@
 #include "tachyon/crypto/commitments/kzg/shplonk.h"
 
 #include <memory>
+#include <string>
 
 #include "gtest/gtest.h"
 
+#include "tachyon/base/containers/container_util.h"
+#include "tachyon/base/json/json.h"
 #include "tachyon/crypto/transcripts/simple_transcript.h"
 #include "tachyon/math/elliptic_curves/bn/bn254/bn254.h"
 #include "tachyon/math/elliptic_curves/bn/bn254/g1.h"
 #include "tachyon/math/elliptic_curves/bn/bn254/g2.h"
 #include "tachyon/math/polynomials/univariate/univariate_evaluation_domain_factory.h"
 
-namespace tachyon::crypto {
+namespace tachyon {  // namespace base
+namespace crypto {
 
+template <typename Poly, typename Commitment>
+struct OwnedPolynomialOpening {
+  Poly poly;
+  Commitment commitment;
+  typename Poly::Point point;
+  typename Poly::Field opening;
+};
+
+template <typename Poly, typename Commitment>
+struct OwnedPolynomialOpenings {
+  std::vector<OwnedPolynomialOpening<Poly, Commitment>> prover_openings;
+
+  template <typename PCS>
+  void Validate(const PCS& pcs) const {
+    for (const OwnedPolynomialOpening<Poly, Commitment>& owned_opening :
+         prover_openings) {
+      CHECK(owned_opening.poly.Evaluate(owned_opening.point) ==
+            owned_opening.opening);
+      Commitment commitment;
+      CHECK(pcs.Commit(owned_opening.poly, &commitment));
+      CHECK(commitment == owned_opening.commitment);
+    }
+  }
+
+  std::vector<PolynomialOpening<Poly>> CreateProverOpenings() const {
+    return base::Map(
+        prover_openings,
+        [](const OwnedPolynomialOpening<Poly, Commitment>& owned_opening) {
+          return PolynomialOpening<Poly>(
+              base::ShallowRef<const Poly>(&owned_opening.poly),
+              base::DeepRef<const typename Poly::Point>(&owned_opening.point),
+              owned_opening.opening);
+        });
+  }
+
+  std::vector<PolynomialOpening<Poly, Commitment>> CreateVerifyingOpenings()
```

Comment:
`CreateProvingOpening()` and `CreateVerifyingOpening()` doesn't correspond to each other.

```suggestion
  std::vector<PolynomialOpening<Poly, Commitment>> CreateVerifierOpenings()
```